### PR TITLE
Defining system/course/instance announcements abilities

### PR DESCRIPTION
**Pull request by allenwq**

### CODE REVIEW

**Comment by lowjoel** [comment]
Same thing, fix the require

File: app/models/components/ability_host.rb [hunk 1]
```diff
@@ -83,6 +83,17 @@ def currently_valid_hashes
         }
       ]
     end
+
+    # Returns a condition which will return started rows(start_at before current time) when
+    # ORed together in a database query. Reverse-merge this with your conditions to obtain the
+    # set of already started rows in the table.
+    #
+    # @return [Hash] The hash condition.
+    def already_started_hash
```
Comment (lowjoel) [inline app/models/components/ability_host.rb]: Isn't this also time bounded?

File: spec/features/course/announcement_management_spec.rb [hunk 2]
```diff
@@ -81,5 +84,23 @@
                                       text: I18n.t('course.announcements.destroy.success'))
       end
     end
+
+    context 'As an Course Student' do
+      let(:user) { create(:course_student, :approved, course: course).user }
+
+      scenario 'I can see the started announcements' do
```
Comment (lowjoel) [inline spec/features/course/announcement_management_spec.rb]: valid announcements?

File: spec/features/system/admin/instance/instance_announcement_management_spec.rb [hunk 3]
```diff
@@ -6,13 +6,13 @@
...
 
   with_tenant(:instance) do
-    let(:user) { create(:administrator) }
-
     before do
       login_as(user, scope: :user)
     end
 
-    context 'As a system administrator' do
+    context 'As a instance administrator' do
```
Comment (lowjoel) [inline spec/features/system/admin/instance/instance_announcement_management_spec.rb]: as an

File: spec/models/generic_announcement_ability_spec.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
+require 'rails_helper'
+
+RSpec.describe GenericAnnouncement do
```
Comment (lowjoel) [inline spec/models/generic_announcement_ability_spec.rb]: Do we directly use generic announcements in the code base?

File: app/models/components/ability_host.rb (hunk 1 above)
Comment (allenwq) [inline app/models/components/ability_host.rb]: Yes, but these are the already started ones ( including those expired rows )

File: spec/models/generic_announcement_ability_spec.rb (hunk 4 above)
Comment (allenwq) [inline spec/models/generic_announcement_ability_spec.rb]: Yes, in some places: https://github.com/Coursemology/coursemology2/blob/9f23d6a2eec4dd59b607f6c7b7e7d0a43cc88eda/app/controllers/concerns/application_announcements_concern.rb#L5

File: app/models/components/ability_host.rb (hunk 1 above)
Comment (lowjoel) [inline app/models/components/ability_host.rb]: Why do we want expired rows?

Comment (allenwq) [inline app/models/components/ability_host.rb]: Even announcements expired, students still can see them in index view right ?

Comment (lowjoel) [inline app/models/components/ability_host.rb]: Then what's the purpose of expiring announcements?

Comment (allenwq) [inline app/models/components/ability_host.rb]: In v1 the expired ones do not show in course home page(the page with activities).

And in v2 it mostly just a indicator now ( don't have course home page yet)?  I mean the we have the calendar icon and a different css styles for expired ones.

Comment (allenwq) [inline app/models/components/ability_host.rb]: I think it makes sense to show the old announcements when user intentionally visits the index page, but we do not actively display/notify the expired ones.

Comment (lowjoel) [inline app/models/components/ability_host.rb]: Then that has nothing to do with permissions.

Comment (lowjoel) [inline app/models/components/ability_host.rb]: That's a filter, and filter is to be applied after permissions.

Comment (lowjoel) [inline app/models/components/ability_host.rb]: So ok I guess this makes sense...

File: spec/features/system/admin/instance/instance_announcement_management_spec.rb (hunk 3 above)
Comment (lowjoel) [inline spec/features/system/admin/instance/instance_announcement_management_spec.rb]: @allenwq just fix this please.

**Comment by allenwq** [comment]
@lowjoel Updated, the announcements pagination specs get random failure very frequently now, I'll fix it in another PR.

**Pull request merged**